src/resampler.rs

```diff
@@ -0,0 +1,259 @@
+// License: MIT
+// Copyright © 2024 Frequenz Energy-as-a-Service GmbH
+
+use chrono::{DateTime, TimeDelta, Utc};
+use log::warn;
+use num_traits::{Float, FromPrimitive};
+use std::fmt::Debug;
+use std::ops::Div;
+
+use itertools::Itertools;
+
+/// The DataType trait represents the data type of the samples.
+pub trait DataType:
+    Div<Output = Self> + std::iter::Sum + FromPrimitive + Float + Default + Debug
+{
+}
+impl DataType for f32 {}
+impl DataType for f64 {}
+
+/// The Sample trait represents a single sample in a time series.
+pub trait Sample: Clone + Debug + Copy + Default {
+    type Value;
+    fn new(timestamp: DateTime<Utc>, value: Option<Self::Value>) -> Self;
+    fn timestamp(&self) -> DateTime<Utc>;
+    fn value(&self) -> Option<Self::Value>;
+}
+
+/// The ResamplingFunction enum represents the different resampling functions that can be used to resample a channel.
+#[derive(Clone)]
+pub enum ResamplingFunction<'a, T: DataType, S: Sample<Value = T>> {
+    /// Calculates the average of all samples in the time step (ignoring None values)
+    Average,
+    /// Calculates the sum of all samples in the time step (ignoring None values)
+    Sum,
+    /// Calculates the maximum value of all samples in the time step (ignoring None values)
+    Max,
+    /// Calculates the minimum value of all samples in the time step (ignoring None values)
+    Min,
+    /// Uses the last sample in the time step. If the last sample is None, the resampling function will return None.
+    Last,
+    /// Counts the number of samples in the time step (ignoring None values)
+    Count,
+    /// A custom resampling function that takes a closure that takes a slice of samples and returns an optional value.
+    Custom(&'a dyn Fn(&[&S]) -> Option<T>),
```

Comment:
I'd think `Box<dyn FnMut...>` is the most flexible.  It would mean we can't `Clone` anymore, but do we really need clone?

`ResamplingFunction` will be a member of the `Resampler`, and there will always be a mutable reference to call `Box::as_mut`, before you apply the function.

Reply:
`FnMut` is more flexible because you can also include immutable functions. But with `Fn`, you can only specify callables that don't have a mutable state. I think mutable states would be helpful for custom resampler functions.